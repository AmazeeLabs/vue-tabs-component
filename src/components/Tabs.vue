<template>
    <div>
        <div class="tabs">
            <ul>
                <li v-for="tab in tabs" :class="{ 'is-active': tab.isActive }">
                    <a :href="tab.hash" @click="selectTab(tab)">
                        <div v-html="tab.header"></div>
                    </a>
                </li>
            </ul>
        </div>
        <div class="tabs-details">
            <slot></slot>
        </div>
    </div>
</template>

<script>
    import Tab from './Tab.vue';

    export default {
        components: {
            Tab,
        },

        props: {
            cacheLifetime: { default: 5 },
        },

        data() {
            return {
                tabs: [],
                activeTabHash: '',
            };
        },

        created() {
            this.tabs = this.$children;
        },

        mounted() {
            const tabWithHash = this.findTab(window.location.hash);

            if (tabWithHash) {
                this.selectTab(tabWithHash);
                return;
            }

            const previousSelectedTab = this.retrieveSelectedTab();

            if (previousSelectedTab) {
                this.selectTab(previousSelectedTab);
                return;
            }

            if (this.tabs.length) {
                this.selectTab(this.tabs[0]);
            }
        },

        methods: {
            selectTab(selectedTab) {
                this.tabs.forEach(tab => {
                    tab.isActive = (tab.hash === selectedTab.hash);
                });

                this.$emit('changed', { tab: selectedTab });

                this.activeTabHash = selectedTab.hash;

                this.rememberSelectedTab(selectedTab);
            },

            findTab(hash) {
                return this.tabs.find(tab => tab.hash === hash);
            },

            rememberSelectedTab(tab) {
                const cache = { hash: tab.hash, expires: this.addMinutes(new Date(), this.cacheLifetime) };

                localStorage.setItem(this.determineLocalStorageKey(), JSON.stringify(cache));
            },

            retrieveSelectedTab() {
                let cache = localStorage.getItem(this.determineLocalStorageKey());

                if (!cache) {
                    return;
                }

                cache = JSON.parse(cache);

                const expiryDate = new Date(cache.expires);

                if (expiryDate < new Date()) {
                    return;
                }

                return this.findTab(cache.hash);
            },

            determineLocalStorageKey() {
                return `vue-tabs-component.cache.${window.location.host}${window.location.pathname}`;
            },

            addMinutes(date, minutes) {
                return new Date(date.getTime() + minutes * 60000);
            },
        },
    };
</script>
