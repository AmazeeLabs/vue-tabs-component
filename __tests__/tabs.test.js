import { beforeEachHooks, afterEachHooks, mount } from 'vue-unit';

import Tabs from '../src/components/Tabs.vue';

describe('save-state', () => {

    beforeEach(() => {
        beforeEachHooks();
    });

    afterEach(() => {
        afterEachHooks();
    });

    it('can mount tabs', () => {
        mount(Tabs);
    });
});