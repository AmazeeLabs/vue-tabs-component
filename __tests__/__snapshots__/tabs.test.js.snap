// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue-tabs-component can accept a prefix and a suffix for the name 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabsComponent\\"><nav class=\\"tabsComponent-tabs\\"><ul><li class=\\"tabsComponent-tab is-active\\"><a href=\\"#first-tab\\"><div>prefixFirst tabsuffix</div></a></li></ul></nav> <section class=\\"tabsComponent-panels\\"><div style=\\"\\">
                        First tab content
                    </div></section></div></div>
        "
`;

exports[`vue-tabs-component can mount tabs 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabsComponent\\"><nav class=\\"tabsComponent-tabs\\"><ul><li class=\\"tabsComponent-tab is-active\\"><a href=\\"#first-tab\\"><div>First tab</div></a></li><li class=\\"tabsComponent-tab\\"><a href=\\"#second-tab\\"><div>Second tab</div></a></li><li class=\\"tabsComponent-tab\\"><a href=\\"#third-tab\\"><div>Third tab</div></a></li></ul></nav> <section class=\\"tabsComponent-panels\\"><div style=\\"\\">
                        First tab content
                    </div> <div style=\\"display: none;\\">
                        Second tab content
                    </div> <div style=\\"display: none;\\">
                        Third tab content
                    </div></section></div></div>
        "
`;

exports[`vue-tabs-component the life time of the cache can be set 1`] = `
Object {
  "vue-tabs-component.cache.blank": "{\\"value\\":\\"#first-tab\\",\\"expires\\":\\"2017-01-01T00:10:00.000Z\\"}",
}
`;

exports[`vue-tabs-component writes the hash of the last opened tab in local storage 1`] = `
Object {
  "vue-tabs-component.cache.blank": "{\\"value\\":\\"#third-tab\\",\\"expires\\":\\"2017-01-01T00:05:00.000Z\\"}",
}
`;
